# Import Libraries

In [2]:
import pandas as pd
from selenium import webdriver
from bs4 import BeautifulSoup
import time

# Extract

## Web Scraping

In [11]:
# Variabel untuk menyimpan data
nama = []
harga = []
diskon = []
jumlah_terjual = []
kota = []

for page in range (1, 4):
    # Buat driver
    driver = webdriver.Chrome()

    # Akses website
    url = f'https://www.lazada.co.id/catalog/?from=hp_categories&page={page}&q=Peralatan%20Dapur%20Kecil&service=all_channel&spm=a211g0.flashsale.cate_8.2.3ccf6nNb6nNbxr&src=all_channel'
    driver.get(url)

    # Tunggu website selesai load
    time.sleep(5)

    # Fetch and clean HTML
    html = driver.page_source
    soup = BeautifulSoup(html, 'html.parser')

    # Mangambil semua data dalam boxes
    boxes = soup.find_all('div', {'class': 'buTCk'})
    # Mengambil data berdasarkan kategori
    for box in boxes:
        # Kategori nama
        try:
            box_nama = box.find('div', {'class': 'RfADt'})
            nama.append(box_nama.get_text())
        except:
            nama.append(None)
        # Kategori harga
        try:
            box_harga = box.find('span', {'class': 'ooOxS'})
            harga.append(box_harga.get_text())
        except:
            harga.append(None)
        # Kategori diskon
        try:
            box_diskon = box.find('span', {'class': 'IcOsH'})
            diskon.append(box_diskon.get_text())
        except:
            diskon.append(None)
        # Kategori jumlah terjual
        try:
            box_terjual = box.find('span', {'class': '_1cEkb'})
            jumlah_terjual.append(box_terjual.get_text())
        except:
            jumlah_terjual.append(None)
        # Kategori kota asal
        try:
            box_kota = box.find('span', {'class': 'oa6ri'})
            kota.append(box_kota.get_text())
        except:
            kota.append(None)

    # Matikan webdriver
    driver.quit()

## Create DataFrame

In [12]:
# Buat DataFrame kosong
df = pd.DataFrame()

# Masukkan lists ke dalam DataFrame
df['nama_produk'] = nama
df['harga_produk'] = harga
df['diskon'] = diskon
df['banyaknya_terjual'] = jumlah_terjual
df['kota_toko'] = kota

# Cek DataFrame
df

,nama_produk,harga_produk,diskon,banyaknya_terjual,kota_toko
0,Pisau set Pisau Guntung Dapur 6In1 AntiLengket...,"Rp35,150",52% Off,882 sold,Kab. Tangerang
1,Spatula Peralatan Masak Sutil Silikon Non Stic...,"Rp21,900",29% Off,11.7K sold,Kota Depok
2,[Best Seller] Samono Chopper 2L Stainless Mult...,"Rp180,999",64% Off,1.5K sold,Kab. Tangerang
3,ANQI SHOP Chopper Blender daging penggiling Me...,"Rp79,126",74% Off,6.5K sold,Kab. Tangerang
4,Blender Portable Alat Penggiling Pencincang Bu...,"Rp26,432",47% Off,168 sold,Kab. Tangerang
...,...,...,...,...,...
115,WDN BISA COD GRATIS ONGKIR Blender Portable Ca...,Rp63.000,37% Off,749 sold,Kota Depok
116,Chopper Mini Blender Multifungsi Portable Elek...,Rp25.920,60% Off,2.4K sold,Kota Surabaya
117,Gum Store Blender 2L Daging Serbaguna Stainles...,Rp81.900,37% Off,153 sold,Kab. Jember
118,Blender Mini Serbaguna Blender Capsule 4 Mata ...,Rp48.999,48% Off,4.2K sold,Kota Jakarta Barat


## Export Raw Data

In [ ]:
df.to_csv('data/scraped_products.csv', index=False)

# Transform

## Data Preparation

### Load data

In [ ]:
df = pd.read_csv('data/scraped_products.csv')

### Explore Data

In [10]:
# Meanmpilkan DataFrame
df

,nama_produk,harga_produk,diskon,banyaknya_terjual,kota_toko
0,Pisau set Pisau Guntung Dapur 6In1 AntiLengket...,"Rp35,150",52% Off,882 sold,Kab. Tangerang
1,Spatula Peralatan Masak Sutil Silikon Non Stic...,"Rp21,900",29% Off,11.7K sold,Kota Depok
2,[Best Seller] Samono Chopper 2L Stainless Mult...,"Rp180,999",64% Off,1.5K sold,Kab. Tangerang
3,ANQI SHOP Chopper Blender daging penggiling Me...,"Rp79,126",74% Off,6.5K sold,Kab. Tangerang
4,Blender Portable Alat Penggiling Pencincang Bu...,"Rp26,432",47% Off,168 sold,Kab. Tangerang
...,...,...,...,...,...
115,WDN BISA COD GRATIS ONGKIR Blender Portable Ca...,Rp63.000,37% Off,749 sold,Kota Depok
116,Chopper Mini Blender Multifungsi Portable Elek...,Rp25.920,60% Off,2.4K sold,Kota Surabaya
117,Gum Store Blender 2L Daging Serbaguna Stainles...,Rp81.900,37% Off,153 sold,Kab. Jember
118,Blender Mini Serbaguna Blender Capsule 4 Mata ...,Rp48.999,48% Off,4.2K sold,Kota Jakarta Barat


In [11]:
# Menampilkan informasi DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   nama_produk        120 non-null    object
 1   harga_produk       120 non-null    object
 2   diskon             112 non-null    object
 3   banyaknya_terjual  119 non-null    object
 4   kota_toko          120 non-null    object
dtypes: object(5)
memory usage: 4.8+ KB


#### Informasi Struktur Data
Dataset memiliki 120 rows dan 5 columns.

#### Struktur Data
##### 1. nama_produk
- Deskripsi: Nama dari produk yang dijual
- Non-null count: 120
- Tipe data: object
- Data lengkap

##### 2. harga_produk
- Deskripsi: Harga produk uang dijual
- Non-null count: 120
- Tipe data: object
- Data lengkap
- Tipe data tidak sesuai dan dibutuhkan pengubahan data ke numerik

##### 3. diskon
- Deskripsi: Informasi mengenai diskon dari produk
- Non-null count: 112
- Tipe data: object
- Terdapat 8 data kosong untuk produk yang tidak diskon
- Tipe data dapat diubah ke numerik untuk analisis

##### 4. banyaknya_terjual
- Deskripsi: Jumlah penjualan produk
- Non-null count: 119
- Tipe data: object
- Terdapat 1 data kosong untuk produk yang tidak terjual
- Tipe data tidak sesuai dan dibutuhkan pengubahan data ke numerik

##### 5. kota_toko
- Deskripsi: Lokasi penjual atau asal pengiriman
- Non-null count: 120
- Tipe data: object
- Data lengkap

#### Kesimpulan Data
Konversi harga_produk, diskon dan banyaknya_terjual ke numerik supaya bisa dilakukan analisis.

### Clean Data

#### Pengubahan data pada kolom harga_produk

In [12]:
# Hapus karakter yang bukan angka
df['harga_produk'] = df['harga_produk'].str.replace('Rp', '')
df['harga_produk'] = df['harga_produk'].str.replace(',', '')
df['harga_produk'] = df['harga_produk'].str.replace('.', '')

# Mengubah tipe data ke numerik
df['harga_produk'] = df['harga_produk'].astype(int)

# Cek tipe data harga_produk
df['harga_produk'].dtypes

dtype('int64')

#### Pengubahan data pada kolom diskon

In [13]:
# Hapus karakter yang bukan angka
df['diskon'] = df['diskon'].str.replace('% Off', '')

# Mengubah NaN menjadi '0'
df['diskon'] = df['diskon'].fillna('0')

# Mengubah tipe data ke numerik
df['diskon'] = df['diskon'].astype(int)

# Cek tipe data diskon
df['diskon'].dtypes

dtype('int64')

#### Pengubahan data pada kolom banyaknya_terjual

In [14]:
# Hapus karakter yang bukan angka
df['banyaknya_terjual'] = df['banyaknya_terjual'].str.replace(' sold', '')
df['banyaknya_terjual'] = df['banyaknya_terjual'].str.replace('.', '')
df['banyaknya_terjual'] = df['banyaknya_terjual'].str.replace('K', '00')

# Mengubah NaN menjadi '0'
df['banyaknya_terjual'] = df['banyaknya_terjual'].fillna('0')

# Mengubah tipe data ke numerik
df['banyaknya_terjual'] = df['banyaknya_terjual'].astype(int)

# Cek tipe data banyaknya_terjual
df['banyaknya_terjual'].dtypes

dtype('int64')

#### Validasi Data

In [15]:
# Menampilkan DataFrame
df

,nama_produk,harga_produk,diskon,banyaknya_terjual,kota_toko
0,Pisau set Pisau Guntung Dapur 6In1 AntiLengket...,35150,52,882,Kab. Tangerang
1,Spatula Peralatan Masak Sutil Silikon Non Stic...,21900,29,11700,Kota Depok
2,[Best Seller] Samono Chopper 2L Stainless Mult...,180999,64,1500,Kab. Tangerang
3,ANQI SHOP Chopper Blender daging penggiling Me...,79126,74,6500,Kab. Tangerang
4,Blender Portable Alat Penggiling Pencincang Bu...,26432,47,168,Kab. Tangerang
...,...,...,...,...,...
115,WDN BISA COD GRATIS ONGKIR Blender Portable Ca...,63000,37,749,Kota Depok
116,Chopper Mini Blender Multifungsi Portable Elek...,25920,60,2400,Kota Surabaya
117,Gum Store Blender 2L Daging Serbaguna Stainles...,81900,37,153,Kab. Jember
118,Blender Mini Serbaguna Blender Capsule 4 Mata ...,48999,48,4200,Kota Jakarta Barat


In [16]:
# Menampilkan informasi DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   nama_produk        120 non-null    object
 1   harga_produk       120 non-null    int64 
 2   diskon             120 non-null    int64 
 3   banyaknya_terjual  120 non-null    int64 
 4   kota_toko          120 non-null    object
dtypes: int64(3), object(2)
memory usage: 4.8+ KB


In [17]:
# Menampilkan informasi statistik deskriptif data numerik
df.describe()

,harga_produk,diskon,banyaknya_terjual
count,120.000000,120.000000,120.000000
mean,38970.908333,45.933333,5807.716667
std,33427.128099,23.859439,13769.657606
min,4000.000000,0.000000,0.000000
25%,16974.250000,25.000000,179.250000
50%,29453.000000,49.000000,822.500000
75%,52452.000000,64.250000,4050.000000
max,250000.000000,83.000000,83000.000000


## Export Clean Data

In [ ]:
df.to_csv('data/clean.csv', index=False)